src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+                        } else {
+                            PartialOrder::less_equal(&resume_upper, &persist_upper)
+                        };
+
+                        // Sort by ts and only present updates that are not
+                        // beyond the resume upper.
+                        persist_stash.sort_unstable_by(|a, b| {
+                            let (_, _, ts1, _) = a;
+                            let (_, _, ts2, _) = b;
+                            Ord::cmp(
+                                ts1,
+                                ts2,
+                            )
+                        });
+
+                        // Find the prefix that we can ingest.
+                        let idx = persist_stash.partition_point(|(_, _, ts, _)| {
+                            let first_source_snapshot = ts == &G::Timestamp::minimum();
+                            let rehydration_snapshot = !resume_upper.less_equal(ts);
+                            first_source_snapshot || rehydration_snapshot
+                        });
+
+                        tracing::debug!(persist_stash = %persist_stash.len(), %idx, %last_snapshot_chunk, ?resume_upper, ?persist_upper, "ingesting persist snapshot chunk");
+
+                        // Check for errors in the prefix that we will ingest.
+                        for (_, value, _ts, diff) in persist_stash.iter_mut().take(idx) {
+                            if let Err(UpsertError::Value(ref mut err)) = value {
+                                // If we receive a legacy error in the snapshot we will keep a note of it but
+                                // insert a non-legacy error in our state. This is so that if this error is
+                                // ever retracted we will correctly retract the non-legacy version because by
+                                // that time we will have emitted the error correction, which happens before
+                                // processing any of the new source input.
+                                if err.is_legacy_dont_touch_it {
+                                    legacy_errors_to_correct.push((err.clone(), diff.clone()));
+                                    err.is_legacy_dont_touch_it = false;
+                                }
+                            }
+                        }
+
+
+                        hydrating = !last_snapshot_chunk;
+
+                        match state
+                            .consolidate_snapshot_chunk(
+                                persist_stash
+                                    .drain(..idx)
+                                    .map(|(key, val, _ts, diff)| (key, val, diff)),
+                                last_snapshot_chunk,
+                            )
+                            .await
+                        {
+                            Ok(_) => {
+                                let _ = snapshot_cap.downgrade(persist_upper.iter());
```

Comment:
Hm, why do we do a `let _ = ` here? Downgrade doesn't return anything

Reply:
sorry, that's left over from when I was debugging a thing with `try_downgrade`